Sources/SwiftGRPCNIO/HTTP1ToRawGRPCClientCodec.swift

```diff
@@ -0,0 +1,153 @@
+      } catch {
+        ctx.fireErrorCaught(error)
+        state = .ignore
+      }
+
+    case .end(let trailers):
+      state = processTrailers(ctx: ctx, trailers: trailers)
+    }
+  }
+
+  /// Forwards the headers from the request head to the next handler.
+  ///
+  /// - note: Requires the `.expectingHeaders` state.
+  private func processHead(ctx: ChannelHandlerContext, head: HTTPResponseHead) -> State {
+    guard case .expectingHeaders = state
+      else { preconditionFailure("received headers while in state \(state)") }
+
+    if let encodingType = head.headers["grpc-encoding"].first {
+      inboundCompression = CompressionMechanism(rawValue: encodingType) ?? .unknown
+    }
+
+    ctx.fireChannelRead(wrapInboundOut(.headers(head.headers)))
+    return .expectingBodyOrTrailers
+  }
+
+  /// Processes the given buffer; if a complete message is read then it is forwarded to the
+  /// next channel handler.
+  ///
+  /// - note: Requires the `.expectingBodyOrTrailers` state.
+  private func processBody(ctx: ChannelHandlerContext, messageBuffer: inout ByteBuffer) throws -> State {
+    guard case .expectingBodyOrTrailers = state
+      else { preconditionFailure("received body while in state \(state)") }
+
+    while messageBuffer.readableBytes > 0 {
+      if let message = try self.messageReader.read(messageBuffer: &messageBuffer, compression: inboundCompression) {
+        ctx.fireChannelRead(wrapInboundOut(.message(message)))
+      }
+    }
+
+    return .expectingBodyOrTrailers
+  }
+
+  /// Forwards a `GRPCStatus` to the next handler. The status and message are extracted
+  /// from the trailers if they exist; the `.unknown` status code is used if no status exists.
+  private func processTrailers(ctx: ChannelHandlerContext, trailers: HTTPHeaders?) -> State {
+    guard case .expectingBodyOrTrailers = state
+      else { preconditionFailure("received trailers while in state \(state)") }
+
+    let statusCode = trailers?["grpc-status"].first
+      .flatMap { Int($0) }
+      .flatMap { StatusCode(rawValue: $0) }
+    let statusMessage = trailers?["grpc-message"].first
+
+    ctx.fireChannelRead(wrapInboundOut(.status(GRPCStatus(code: statusCode ?? .unknown, message: statusMessage))))
+    return .ignore
+  }
+}
+
+extension HTTP1ToRawGRPCClientCodec: ChannelOutboundHandler {
+  public typealias OutboundIn = RawGRPCClientRequestPart
+  public typealias OutboundOut = HTTPClientRequestPart
+
+  public func write(ctx: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    switch unwrapOutboundIn(data) {
+    case .head(let requestHead):
+      ctx.write(wrapOutboundOut(.head(requestHead)), promise: promise)
+
+    case .message(let message):
+      do {
+        let request = try messageWriter.write(allocator: ctx.channel.allocator, compression: .none, message: message)
+        ctx.write(wrapOutboundOut(.body(.byteBuffer(request))), promise: promise)
+      } catch {
+        ctx.fireErrorCaught(error)
```

Comment:
Switch to state `.ignore` here as well? And/or fail the promise?

Reply:
I removed the error throwing from the writer, since we know ahead of time that we'll be using a supported compression mechanism.